When using PinPad, show timeout warning to user

olegprokofjev opened the pull request:
IB-4807:
   When user is using PinPad for either loading picture or changing PIN1
   or PIN2 or PUK, show the warning message on timeout (also if user
   presses Cancel button).
   It will work the similar way as in qdigidoc (which is using as card
   driver the MS Cryptography API).

Signed-off-by: Oleg Prokofjev <oleg@aktors.ee>

olegprokofjev committed:
When using PinPad, show timeout warning to user

IB-4807:
   When user is using PinPad for either loading picture or changing PIN1
   or PIN2 or PUK, show the warning message on timeout (also if user
   presses Cancel button).
   It will work the similar way as in qdigidoc (which is using as card
   driver the MS Cryptography API).

Signed-off-by: Oleg Prokofjev <oleg@aktors.ee>

metsma commented:
Why not show progress bar like DigiDoc does?

olegprokofjev commented:
qesteidutil.exe (seems tera.exe as well) is using Common\QPCSC.cpp driver.
Partly it shows the own made progress bar (Load Picture, @eesti.ee email, Mobile-ID, Update certificate, ...), partly it does not (Changing PIN/PUK codes). I do not know the background - why it was done so.

The qdigidocclient.exe is using MS functionality (like NCryptSignHash) which shows the progress bar on language, defined in computer's Region & Language settings (Win).

metsma commented:
But DigiDoc on OSX/Linux shows custom Progress Bar

olegprokofjev commented:
Yes, for OSX/Linux the qDigiDoc is using different functionality.
I can try to show the progress bar on change of PIN/PUK codes in qEstEidUtil although it was not requested in ticket.

olegprokofjev committed:
Show Progress Bar on change PIN/PUK with PinPad

IB-4807:
   Show Progress Bar on change PIN/PUK with PinPad

Signed-off-by: Oleg Prokofjev <oleg@aktors.ee>

metsma commented on src/MainWindow.cpp:
```diff
@@ -196,7 +196,17 @@ bool MainWindowPrivate::validateCardError( QSmartCardData::PinType type, int fla
+			case QSmartCardData::PukType: showWarning( tr("PUK timeout") ); break;
+			}
+		}
+		break;
```
why additional dialog? when progress bar disappears, it should already indicate that its expired

olegprokofjev replied:
It was a main requirement of IB-4807. Same way it works in qDigidoc3 client. I suppose that both applications should work the same way. Also user can cancel operation (progress bar), but it continue to work in background until time limit and then the timeout message is shown, which is OK to my mind. Tells user that she/he can continue using application.

metsma replied:
DigiDoc3 client does not show message on OSX/Linux

olegprokofjev replied:
Correct. I will remove 'timeout' message dialog for OSX/Linux.

metsma replied:
Same is also on windows, non estonian cards

metsma commented on src/QSmartCard.cpp:
```diff
@@ -233,20 +233,33 @@ QSmartCard::~QSmartCard()
 {
 	QMutexLocker locker(&d->m);
-	QSharedPointer<QPCSCReader> reader(d->connect(d->t.reader()));
+ 	QSharedPointer<QPCSCReader> reader(d->connect(d->t.reader()));
```
Whitespace change?

olegprokofjev committed:
IB-4807: Fix

   Show 'Timeout ...' message box only for Windows (not OSX/Linux) users,
   having Estonian ID cards.

Signed-off-by: Oleg Prokofjev <oleg@aktors.ee>

olegprokofjev committed:
Merge branch 'master' into IB-4807

Signed-off-by: Oleg Prokofjev oleg@aktors.ee

metsma commented:
qesteidutil is now feature complete and no new development is made.
[DigiDoc4-Client](/open-eid/DigiDoc4-Client) supersedes qdigidoc
